tcp.c

```diff
@@ -418,91 +418,52 @@ tcp_tls_get_server_pubkey(STREAM s)
 
 /* Establish a connection on the TCP layer */
 RD_BOOL
-tcp_connect(char *server)
+tcp_connect(struct addrinfo *ai)
 {
-	socklen_t option_len;
+	int i, res;
 	uint32 option_value;
-	int i;
-
-#ifdef IPv6
-
-	int n;
-	struct addrinfo hints, *res, *ressave;
-	char tcp_port_rdp_s[10];
-
-	snprintf(tcp_port_rdp_s, 10, "%d", g_tcp_port_rdp);
-
-	memset(&hints, 0, sizeof(struct addrinfo));
-	hints.ai_family = AF_UNSPEC;
-	hints.ai_socktype = SOCK_STREAM;
-
-	if ((n = getaddrinfo(server, tcp_port_rdp_s, &hints, &res)))
-	{
-		logger(Core, Error, "tcp_connect(), getaddrinfo() failed: %s", gai_strerror(n));
-		return False;
-	}
+	char hostname[64];
+	struct addrinfo *pai;
+	socklen_t option_len;
 
-	ressave = res;
+	/* connect to host */
+	pai = ai;
 	g_sock = -1;
-	while (res)
+	while (pai)
 	{
-		g_sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
-		if (!(g_sock < 0))
+		g_sock = socket(pai->ai_family, pai->ai_socktype, pai->ai_protocol);
+		if (g_sock > 0)
 		{
-			if (connect(g_sock, res->ai_addr, res->ai_addrlen) == 0)
+			/* try connect to server */
+			res = connect(g_sock, pai->ai_addr, pai->ai_addrlen);
+			if (res == 0)
+			{
+				/* We are connect, lets break out of connect loop */
 				break;
+			}
+
+			/* connection failed, lets try next */
 			TCP_CLOSE(g_sock);
 			g_sock = -1;
 		}
-		res = res->ai_next;
-	}
-	freeaddrinfo(ressave);
-
-	if (g_sock == -1)
-	{
-		logger(Core, Error, "tcp_connect(), unable to connect to %s", server);
-		return False;
-	}
-
-#else /* no IPv6 support */
-
-	struct hostent *nslookup;
-	struct sockaddr_in servaddr;
-
-	if ((nslookup = gethostbyname(server)) != NULL)
-	{
-		memcpy(&servaddr.sin_addr, nslookup->h_addr, sizeof(servaddr.sin_addr));
-	}
-	else if ((servaddr.sin_addr.s_addr = inet_addr(server)) == INADDR_NONE)
-	{
-		logger(Core, Error, "tcp_connect(), unable to resolve host '%s'", server);
-		return False;
+		pai = pai->ai_next;
 	}
 
-	if ((g_sock = socket(AF_INET, SOCK_STREAM, 0)) < 0)
+	if (pai == NULL || g_sock == -1)
 	{
-		logger(Core, Error, "tcp_connect(), socket() failed: %s", TCP_STRERROR);
-		return False;
-	}
-
-	servaddr.sin_family = AF_INET;
-	servaddr.sin_port = htons((uint16) g_tcp_port_rdp);
-
-	if (connect(g_sock, (struct sockaddr *) &servaddr, sizeof(struct sockaddr)) < 0)
-	{
-		if (!g_reconnect_loop)
-			logger(Core, Error, "tcp_connect(), connect() failed: %s", TCP_STRERROR);
-
-		TCP_CLOSE(g_sock);
-		g_sock = -1;
+		logger(Core, Error, "tcp_connect(), unable to connect to %s",
+		       ai->ai_canonname);
 		return False;
 	}
 
-#endif /* IPv6 */
+	getnameinfo(pai->ai_addr,  pai->ai_addrlen, hostname, sizeof(hostname), NULL, 0, NI_NUMERICHOST);
+	logger(Core, Debug, "Connected to %s", hostname);
```

Comment:
I don't see it storing the result anywhere. It needs to remember which address it actually used or we'll get problems with reconnects after redirects.

Reply:
I will fix that, this is also needed for CredSSP to resolve the correct name for the connection in case of other then first entry is used.